File: leetcode/323. Number of Connected Components in an Undirected Graph.py
'''
Given n nodes labeled from 0 to n - 1 and a list of undirected edges (each edge is a pair of nodes), 
write a function to find the number of connected components in an undirected graph.

Example 1:

Input: n = 5 and edges = [[0, 1], [1, 2], [3, 4]]

     0          3
     |          |
     1 --- 2    4 

Output: 2
Example 2:

Input: n = 5 and edges = [[0, 1], [1, 2], [2, 3], [3, 4]]

     0           4
     |           |
     1 --- 2 --- 3

Output:  1
Note:
You can assume that no duplicate edges will appear in edges. Since all edges are undirected, [0, 1] is the same as [1, 0] and thus will not appear together in edges.
'''

#自己想的, time complexity O(V+E), space complexity O(V+E), 92ms
#經典 dfs graph題 => 用bfs 也可以
from collections import defaultdict
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = defaultdict(list)
        for i, j in edges:
            graph[i].append(j)
            graph[j].append(i)
            
        visited = set()
        count = 0
        for i in range(n):
            if i not in visited:
                count += 1
                self.dfs(i, graph, visited)
        return count
    
    def dfs(self, node, graph, visited):
        visited.add(node)
        for neighbor in graph[node]:
            if neighbor not in visited:
                self.dfs(neighbor, graph, visited)


#重寫第二次, time complexity O(n), space complexity O(n)
from collections import defaultdict
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = defaultdict(list)
        for edge in edges:
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])
        visited = set()
        count = 0
        for i in range(n):
            if i not in visited:
                count += 1
                self.dfs(i, graph, visited)
        return count
    
    def dfs(self, i, graph, visited):
        if i in visited:
            return
        visited.add(i)
        for j in graph[i]:
            self.dfs(j, graph, visited)
        return


#重寫第三次, time complexity O(n), space complexity O(n)
from collections import defaultdict
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = defaultdict(list)
        for edge in edges:
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])
        visited = set()
        count = 0
        for i in range(n):
            if i not in visited:
                count += 1
                self.dfs(i, graph, visited)
        return count
    
    def dfs(self, i, graph, visited):
        visited.add(i)
        for neighbor in graph[i]:
            if neighbor not in visited:
                self.dfs(neighbor, graph, visited)

File: leetcode_cpp/323. Number of Connected Components in an Undirected Graph.cpp
// 刷題用這個, time complexity O(V+E), space complexity O(V+E)
class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        unordered_map<int, vector<int>> graph;
        for (auto edge: edges) {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }
        vector<int> visited(n, 0);
        int count {0};
        for (int i=0; i<n; i++) {
            if (visited[i] == 0) {
                count++;
                dfs(i, graph, visited);
            }
        }
        return count;
    }

    void dfs(int i, unordered_map<int, vector<int>>& graph, vector<int>& visited) {
        visited[i] = 1;
        for (auto nxt: graph[i]) {
            if (visited[nxt] == 0) {
                dfs(nxt, graph, visited);
            }
        }
    }
};

// 刷題也可用這個
/*
    Graph of n nodes, given edges array, return # of connected components
    Ex. n = 5, edges = [[0,1],[1,2],[3,4]] -> 2

    Union find, for each edge combine, if already in same set keep traversing
    If not in same set, decrement count by 1, count will store # of components

    Time: O(n)
    Space: O(n)
*/

class Solution {
public:
    int countComponents(int n, vector<vector<int>>& edges) {
        vector<int> parents;
        vector<int> ranks;
        for (int i = 0; i < n; i++) {
            parents.push_back(i);
            ranks.push_back(1);
        }
        
        int result = n;
        for (int i = 0; i < edges.size(); i++) {
            int n1 = edges[i][0];
            int n2 = edges[i][1];
            result -= doUnion(parents, ranks, n1, n2);
        }
        return result;
    }
private:
    int doFind(vector<int>& parents, int n) {
        int p = parents[n];
        while (p != parents[p]) {
            parents[p] = parents[parents[p]];
            p = parents[p];
        }
        return p;
    }
    
    int doUnion(vector<int>& parents, vector<int>& ranks, int n1, int n2) {
        int p1 = doFind(parents, n1);
        int p2 = doFind(parents, n2);
        if (p1 == p2) {
            return 0;
        }
        
        if (ranks[p1] > ranks[p2]) {
            parents[p2] = p1;
            ranks[p1] += ranks[p2];
        } else {
            parents[p1] = p2;
            ranks[p2] += ranks[p1];
        }
        
        return 1;
    }
};

